ops/lib/__init__.py

```diff
@@ -108,53 +113,92 @@ def _find_all_specs(path):
             try:
                 lib_dirs = os.listdir(opslib)
             except OSError:
+                # logger.trace(...) # *lots* of things checked here
                 continue
+            else:
+                logger.debug("Trying %r.", opslib)
             finder = get_importer(opslib)
-            if finder is None or not hasattr(finder, 'find_spec'):
+            if finder is None:
+                logger.debug("Finder for %r is None.", opslib)
+                continue
+            if not hasattr(finder, 'find_spec'):
+                logger.debug("Finder for %r has no find_psec.", opslib)
                 continue
             for lib_dir in lib_dirs:
-                spec = finder.find_spec("{}.opslib.{}".format(top_dir, lib_dir))
+                spec_name = "{}.opslib.{}".format(top_dir, lib_dir)
+                spec = finder.find_spec(spec_name)
                 if spec is None:
+                    logger.debug("No spec for %r.", spec_name)
                     continue
                 if spec.loader is None:
                     # a namespace package; not supported
+                    logger.debug("No loader for %r (probably a namespace package)", spec_name)
                     continue
+
+                logger.debug("Found %r.", spec_name)
                 yield spec
 
 
 # only the first this many lines of a file are looked at for the LIB* constants
 _MAX_LIB_LINES = 99
 
 
+_EXPECTED = {'NAME': str, 'AUTHOR': str, 'API': int, 'PATCH': int}
+
+
+# XXX Names Are Hard™
+class _NAH:
+    """A silly little helper to only work out the difference between
+    what was found and what was expected when logging"""
+
+    def __init__(self, found):
+        self._found = found
+
+    def __str__(self):
+        return ", ".join(set(_EXPECTED) - set(self._found))
+
+
 def _parse_lib(spec):
     if spec.origin is None:
+        # "can't happen"
+        logger.debug("No origin for %r (no idea why; please report)", spec.name)
         return None
 
-    _expected = {'NAME': str, 'AUTHOR': str, 'API': int, 'PATCH': int}
+    logger.debug("Parsing %r:", spec.name)
 
     try:
```

Comment:
Consider a refactor of this logic.

What about if instead of failing at the first detail, you continue until EOF or reaching `_MAX_LIB_LINES`, and then get all gathered variables, and at that point it happens the verifications... if len(libinfo) == len(expected), and if all vars are the indicated type, etc?

This will simplify that block of code (the try/except super big, etc), and has the nice user-facing benefit of reporting all problems at once (not one per run).

Reply:
ok, but not as part of this work